### PR TITLE
Accelerate CI builds substantially

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

**Extra short summary:** Reduced build time for almost all binaries from ~30 min to ~3 min

---

**Longer description:**

Building FTL binaries using `buildx` slowed down our CI to roughly 16 minutes per build. If builds sporadically fail, they are repeated and the build needs half an hour. This is quite some unfortunate situation as we need to wait half an hour after pushing a fix until a user can try it.

This PR changes how Pi-hole builds binaries on the CI. Before, we built (and tested!) all binaries in parallel using GHA runners (`x86_64`). Only once all of them were finished, the binaries were uploaded:

Screenshot-old-CI

The new method has three partially independent branches:

1. We build `riscv64` on GHA runners
   Screenshot from 2024-03-08 09-23-23
2. We build `x86_64` and `i386` in parallel on GHA runners
   Screenshot from 2024-03-08 09-23-26
3. We build all `arm` binaries on **self-hosted ARM64-powered runners**
   Screenshot from 2024-03-08 09-23-29

Once `x86_64`, `i386`, and all the `arm` builds are finished (typically **3 minutes** now), we deploy them. The `riscv64` binaries will take a much longer time (there are self-hosted servers available at this point and building them on `arm64` is even slower than on `x86`) but since they are not needed by the vast majority of users, we accept that there is still some larger delay here.

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
7. I have commented my proposed changes within the code.
8. I am willing to help maintain this change if there are issues with it later.
9. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
10. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -61,132 +61,117 @@ jobs:
...
+          bin_name: ${{ matrix.bin_name }}
+          artifact_name: ${{ matrix.bin_name }}-binary
+          git_branch: ${{ needs.smoke-tests.outputs.GIT_BRANCH }}
+          git_tag: ${{ needs.smoke-tests.outputs.GIT_TAG }}
+          event_name: ${{ github.event_name }}
 
   deploy:
     if: github.event_name != 'pull_request'
-    needs: [smoke-tests, build]
+    needs: [smoke-tests, build-gha-x86, build-self-hosted]
```
Comment (PromoFaux) [inline .github/workflows/build.yml:139]: Just reading over this file again. I'm wondering if `smoke-tests` is `need`ed at the `deploy` stage?

if `build-gha-x86` or `build-self-hosted` are true, then it stands to reason that `smoke-tests` must also be true (else they would never start)

I think I've logicked that correctly? Why isn't logicked a real word?

**Comment by DL6ER** [comment]
Further improvement (direct deployment) on its way... changed back to draft mode and will re-request review later when it's ready.

**Comment by DL6ER** [comment]
We used to have a combined `deploy`ment step as - pre-v6 - we have still been using cross-compilation toolchains and been only been able to test the `x86_64` build but not the `arm` builds. Hence, we wanted to wait for `arm` binary deployment until we had a success message from the CI tests.

With Pi-hole v6, we use `docker buildx` instead of cross-compilation allowing us to test each of the binaries on their own. Hence, we can directly deploy everything which passes the CI tests. If any architecture fails due to build for some arch-specific thing, merging the PR will still be blocked by the failed CI build. The fact that the other binaries have been deployed is harmless as they have been tested okay themselves.

Updated workflow screenshot with immediate deployment:
grafik

resulting in:
grafik

**Comment by DL6ER** [comment]
On its last iteration, it got even simpler:
grafik

**Review by PromoFaux — APPROVED** [review]

**Pull request merged**